### ISSUE TITLE
Allow vector constraints and goals

### ISSUE THREAD

**Issue by SGeeversAtVortech** [Deltares/rtc-tools#1046]
In GitLab by @vreeken on Jun 15, 2018, 10:01

Currently all goal functions are assumed to be scalar, and so are the constraints. It turns out it can be much faster for 10k+ constraints/goals to try to make as few calls to CasADi, to avoid the SWIG/GIL overhead involved with each call. Note that a single expression X1 + X2 + X3 will likely go back and forth at least 3 times.

If the user can vertcat symbols in his expressions, a lot of time can be saved. This only makes sense of course when the expressions are very similar and can actually be vectorized. However, the common structure of expressions for e.g. goals can be easily exploited, resulting in a 100x speedup for 100k (state)goals (see code snippet below).

Note that many places in the code can currently only handle scalar expressions, and some reworking may be needed to allow vector expressions (if at all possible).

* [ ]  Figure out how to check if addition of three elements (like X1 + X2 + X3) will indeed go back and forth 3 times.
* [ ]  Similarly, check if a ca.vertcat on 100k elements will only go back and forth _once_

```python
import casadi as ca
import random
import timeit

n = 100_000

X = [ca.MX.sym("X[%d]" % i) for i in range(n)]

epsilons = random.choices(range(n), k=n)
target_mins = random.choices(range(n), k=n)
mins = random.choices(range(n), k=n)
nominals = random.choices(range(2, 1000), k=n)

# Current approach
lambdas = []

for eps, tmin, min_, nominal in zip(epsilons, target_mins, mins, nominals):
    def _f(X, eps=eps, tmin=tmin, min_=min_, nominal=nominal):
        return (X[eps] * (X[min_] - X[tmin]) + X[eps]) / nominal
    lambdas.append(_f)

all_exprs = [f(X) for f in lambdas]  # Takes about 10-14 seconds to complete!

# Using a vectorized approach, where the same work takes about 70 ms
x_syms = [ca.MX.sym("X[%d]" % i) for i in range(n)]  # 200 ms, but already don

X = ca.vertcat(*x_syms)  # 40 ms
(X[eps] * (X[min_] - X[tmin]) + X[eps]) / nominal  # 30 ms
```

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Dec 27, 2018, 18:15

Solved by 916211bc and a788207e.

**Comment by SGeeversAtVortech** [comment]
In GitLab by @vreeken on Dec 27, 2018, 18:15

closed
